src/kex.c

```diff
@@ -98,6 +98,44 @@
             }                                                              \
     }
 
+/*
+ * Generic Diffie-Hellman computation support.
+ *
+ * DH context should be a _libssh2_bn *.
+ */
+
+#ifndef libssh2_dh_key_pair
```

Comment:
It's confusing using `libssh2_dh_key_pair` both as a function definition and as a `#define`d symbol.  The two things exist in different namespaces.  For example `libssh2_dh_key_pair` defined on line 113 a won't make `#ifndef libssh2_dh_key_pair` true.  But at first glance that looks like what you're trying to do.  I would prefer a separate symbol (something like `#ifndef LIBSSH2_DH_IMPLEMENTATION`).

Though, really, I think the correct solution is to break out these functions into two separate files, where the only one is chosen at config time (like we do for the crypto-backend implementations).  One file would be for for 0S400, the other for everything else.  They would then simply implement the same set of 4 functions, no preprocessor needed.